src/Elastic.CommonSchema.LoggerProvider/ElasticsearchDataProcessor.cs

```diff
@@ -0,0 +1,262 @@
+				var type = assembly.GetName().Name;
+				var versionAttribute = assembly.GetCustomAttributes(false)
+					.OfType<AssemblyInformationalVersionAttribute>()
+					.FirstOrDefault();
+				var version = versionAttribute?.InformationalVersion;
+				_agent = new Agent { Type = type, Version = version };
+			}
+
+			return _agent;
+		}
+
+		internal Host GetHost()
+		{
+			if (_host is null)
+			{
+				// Architecture osArchitecture = RuntimeInformation.OSArchitecture;
+				// if (osDescription.Contains('#'))
+				// {
+				//     int indexOfHash = osDescription.IndexOf('#');
+				//     osDescription = osDescription.Substring(0, Math.Max(0, indexOfHash - 1));
+				// }
+
+				var operatingSystem = new Os
+				{
+					Full = RuntimeInformation.OSDescription,
+					Platform = Environment.OSVersion.Platform.ToString(),
+					Version = Environment.OSVersion.Version.ToString()
+				};
+				_host = new Host
+				{
+					Hostname = Environment.MachineName, Architecture = RuntimeInformation.OSArchitecture.ToString(), Os = operatingSystem
+				};
+			}
+
+			return _host;
+		}
+
+		internal Process GetProcess()
+		{
+			if (_processName == null)
+			{
+				using var process = System.Diagnostics.Process.GetCurrentProcess();
+				_processId = process.Id;
+				_processName = process.ProcessName;
+			}
+
+			var currentThread = Thread.CurrentThread;
+
+			return new Process
+			{
+				Name = _processName,
+				Pid = _processId,
+				Thread = new ProcessThread { Name = currentThread.Name, Id = currentThread.ManagedThreadId }
+			};
+		}
+
+		internal Service GetService()
+		{
+			if (_service is null)
+			{
+				var entryAssembly = Assembly.GetEntryAssembly();
+				var entryAssemblyName = entryAssembly.GetName();
+				var type = entryAssemblyName.Name;
+				var versionAttribute = entryAssembly.GetCustomAttributes(false)
+					.OfType<AssemblyInformationalVersionAttribute>()
+					.FirstOrDefault();
+				var version = versionAttribute?.InformationalVersion ?? entryAssemblyName.Version.ToString();
+				_service = new Service { Type = type, Version = version };
+			}
+
+			return _service;
+		}
+
+		internal int GetSeverity(LogLevel logLevel)
+		{
+			switch (logLevel)
+			{
+				case LogLevel.Critical:
+					return 2;
+				case LogLevel.Error:
+					return 3;
+				case LogLevel.Warning:
+					return 4;
+				case LogLevel.Information:
+					return 6;
+				case LogLevel.Debug:
+				case LogLevel.Trace:
+					return 7;
+				default:
+					return 7;
+			}
+		}
+
+		private void PostEvent(LogEvent logEvent)
+		{
+			var indexTime = logEvent.Timestamp ?? ElasticsearchLoggerProvider.LocalDateTimeProvider();
+			if (_options.IndexOffset.HasValue) indexTime = indexTime.ToOffset(_options.IndexOffset.Value);
+
+			var index = string.Format(_options.Index, indexTime);
+
+			var id = Guid.NewGuid().ToString();
+
+			var localClient = _lowLevelClient;
+			var response = localClient.Index<StringResponse>(index, id, PostData.Serializable(logEvent));
```

Comment:
I think for the first version this is fine but we'd want to look into batching support via `Bulk()` for every N messages or N seconds do a flush. 

It'd be good to move from BlockingCollection to either TPL or channels for this: 

https://michaelscodingspot.com/performance-of-producer-consumer/

TPL is particulary strong in creating blocks out of streams.

Reply:
I have used batching in my previous libraries, but this is only a first version of this one.

One important feature for logging batching, from previous work, is that the first log message should send immediately (single message), then start batching. A user wants to see that first log message immediately, to know the system is working. You don't want to wait seconds / lots of messages before sending the first.

I've used channels before, and they are really good ... but I tried to keep it netstandard20 compliant.

Reply:
Before batching though, I'd suggest the next needed feature I had on my todo list was supporting security/config options for the connection. (next feature)

Reply:
This is for a future version; maybe move to an issue / project card to track?